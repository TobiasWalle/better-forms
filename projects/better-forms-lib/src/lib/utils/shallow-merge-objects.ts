export function shallowMergeObjects<T>(objects: T[]): T {
  return objects.reduce((object, nextObject) => {
    return {
      ...object,
      ...nextObject
    };
  }, {});
}
