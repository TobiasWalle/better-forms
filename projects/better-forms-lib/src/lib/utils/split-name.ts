export function splitName(path: string): string[] {
  return path.split('.').filter(s => s !== '');
}
