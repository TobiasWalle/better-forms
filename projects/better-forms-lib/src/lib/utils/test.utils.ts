import { ControlValueAccessor } from '@angular/forms';
import { Observable } from 'rxjs';
import { take } from 'rxjs/operators';

export function createMockValueAccessor(): ControlValueAccessor {
  return {
    writeValue: jest.fn(),
    registerOnChange: jest.fn(),
    registerOnTouched: jest.fn(),
    setDisabledState: jest.fn(),
  };
}

export async function getLatestValue<T>(observable: Observable<T>): Promise<T> {
  return await observable.pipe(take(1)).toPromise();
}
