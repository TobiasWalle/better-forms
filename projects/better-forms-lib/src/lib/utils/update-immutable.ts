const immutable: any = require('object-path-immutable');

export function setImmutable<T>(original: T, path: string[], newValue: any): T {
  return immutable.set(original, path, newValue);
}

export function getByPath<T>(object: T, path: string[]): any {
  return path.reduce((nextObject: any, key: string) => {
    if (key in nextObject) {
      return nextObject[key];
    }
    throw new Error(`Couldn't find "${path}" on object. "${key}" doesn not exists on ${nextObject}.`);
  }, object);
}

