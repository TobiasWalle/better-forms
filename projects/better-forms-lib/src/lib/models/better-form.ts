import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { getByPath, setImmutable } from '../utils/update.utils';

export interface UpdateInformation {
  /** The path that has been updated */
  path: string[];
  /** The new value of the path */
  newValue: any;
}

export interface BetterFormOptions<T> {
  initialValue: T;
}

export class BetterForm<T> {
  private valueSubject: BehaviorSubject<T>;
  public valueChange: Observable<T>;

  private onUpdateSubject: Subject<UpdateInformation> = new Subject<UpdateInformation>();

  /**
   * Observable that updates every time a value gets updated. It provides information
   * about the update like the updated path or value.
   */
  public onUpdate: Observable<UpdateInformation> = this.onUpdateSubject.asObservable();

  /** The current form value */
  public get value(): T {
    return this.valueSubject.value;
  }

  constructor(options: BetterFormOptions<T>) {
    this.valueSubject = new BehaviorSubject<T>(options.initialValue);
    this.valueChange = this.valueSubject.asObservable();
  }

  /** Override the current form value with a new one */
  public setValue(value: T): void {
    this.triggerUpdate([], value);
  }

  /**
   * Update a specific part of the form value.
   * Use with caution, as this operation is not typesafe yet.
   * @param {string[]} path The path to update
   * @param value The new value for the path.
   */
  public updatePath(path: string[], value: any): void {
    this.triggerUpdate(path, value);
  }

  /**
   * Get a specific value in the form.
   * Use with caution, as this operation is not typesafe yet.
   * @param {string[]} path The path to get the value from.
   * @returns {any} The value behind the path.
   */
  public getByPath(path: string[]): any {
    return getByPath(this.value, path);
  }

  private triggerUpdate(path: string[], newValue: any): void {
    const newFormValue = setImmutable(this.value, path, newValue);
    this.valueSubject.next(newFormValue);
    this.onUpdateSubject.next({ path, newValue });
  }
}
