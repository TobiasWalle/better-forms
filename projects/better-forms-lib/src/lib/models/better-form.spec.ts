import { getLatestValue } from '../utils/test.utils';
import { BetterForm, UpdateInformation } from './better-form';

interface TestValue {
  a?: number;
  b: {
    c: string;
  };
}

describe('BetterForm', () => {
  const initialValue: TestValue = Object.freeze({
    a: 1,
    b: {
      c: 'test'
    }
  });
  let form!: BetterForm<TestValue>;

  beforeEach(() => {
    form = new BetterForm({ initialValue });
  });

  describe('#value', () => {
    it('should return the current value', () => {
      expect(form.value).toEqual(initialValue);
    });
  });

  describe('#setValue', () => {
    it('should set a new value', () => {
      const newValue: TestValue = {
        b: {
          c: 'otherTest'
        }
      };
      form.setValue(newValue);

      expect(form.value).toEqual(newValue);
    });
  });

  describe('#updatePath', () => {
    it('should update a value based on the path', () => {
      form.updatePath(['b', 'c'], 'update');

      expect(form.value).toEqual({
        ...initialValue,
        b: {
          c: 'update'
        }
      });
    });

    it('should not throw an error if the path does not exists', () => {
      expect(() => form.updatePath(['b', 'd'], 'update')).not.toThrowError();
    });
  });

  describe('#getByPath', () => {
    it('should get a value by the path', () => {
      expect(form.getByPath(['b', 'c'])).toBe('test');
    });

    it('should throw an error if the value does not exists', () => {
      expect(() => form.getByPath(['b', 'd'])).toThrowErrorMatchingSnapshot();
    });
  });

  describe('#valueChange', () => {
    async function expectValueChangeToBeTriggeredWith(value: TestValue) {
      expect(await getLatestValue(form.valueChange)).toEqual(value);
    }

    it('should trigger on setValue', async () => {
      await expectValueChangeToBeTriggeredWith(initialValue);

      const newValue: TestValue = { b: { c: 'c' }};
      form.setValue(newValue);

      await expectValueChangeToBeTriggeredWith(newValue);
    });

    it('should trigger on updatePath', async () => {
      await expectValueChangeToBeTriggeredWith(initialValue);

      const newValue: TestValue = {
        ...initialValue,
        b: { c: 'c' }
      };
      form.updatePath(['b', 'c'], 'c');

      await expectValueChangeToBeTriggeredWith(newValue);
    });
  });

  describe('#onUpdate', () => {
    let onUpdateCallback: jest.Mock;

    beforeEach(() => {
      onUpdateCallback = jest.fn();
      form.onUpdate.subscribe(onUpdateCallback);
    });

    it('should trigger on setValue', async () => {
      expect(onUpdateCallback).not.toHaveBeenCalled();

      const newValue: TestValue = { b: { c: 'c' }};
      form.setValue(newValue);

      expect(onUpdateCallback).toHaveBeenCalledWith({
        path: [],
        newValue
      });
    });

    it('should trigger on updatePath', async () => {
      expect(onUpdateCallback).not.toHaveBeenCalled();

      form.updatePath(['b', 'c'], 'c');

      expect(onUpdateCallback).toHaveBeenCalledWith({
        path: ['b', 'c'],
        newValue: 'c'
      });
    });
  });
});
