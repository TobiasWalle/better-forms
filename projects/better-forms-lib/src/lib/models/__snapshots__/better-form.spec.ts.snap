// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetterForm #getByPath should throw an error if the value does not exists 1`] = `
"Couldn't find \\"b,d\\" on object. \\"d\\" does not exists on {
  \\"c\\": \\"test\\"
}."
`;
